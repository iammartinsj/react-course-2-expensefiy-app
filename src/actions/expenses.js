import uuid from 'uuid';

//Add-Expense function
//Destructure Object Parameter
export const addExpense = ({
  description = '',
  note = '',
  amount = 0,
  createdAt = 0
} = {}) => ({
  type:'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

//Remove-Expense function
export const removeExpense = ({id}) => ({
  type:'REMOVE_EXPENSE',
  id
});

//Edit-Expense function
export const editExpense = (id, updates) => ({
  type:'EDIT_EXPENSE',
  id,
  updates
});
