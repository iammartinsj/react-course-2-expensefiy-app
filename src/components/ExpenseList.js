import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {
      props.expenses.length === 0 ?
      (<p>No Expenses</p>) :
      props.expenses.map((expense) => 
        <ExpenseListItem key={expense.id} {...expense} dispatch={props.dispatch} />
      )
    }
  </div>
);

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filter)
});

export default connect(mapStateToProps)(ExpenseList);