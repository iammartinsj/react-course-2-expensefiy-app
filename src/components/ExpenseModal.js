import React from 'react';
import Modal from 'react-modal';

const ExpenseModal = (props) => (
 <Modal
  isOpen={!!props.forRemovalExpense}
  contentLabel="ConfirmDeleteExpense"
  className="modal"
  onRequestClose={props.onClearForRemovalExpense}
 >
   <h2 className="modal__title">Remove Expense</h2>
   <div className="modal__body">
   {
     props.forRemovalExpense && 
     <div>
      <p>Are you sure you want to remove</p> 
      <b>{props.forRemovalExpense}?</b>
     </div>
   }
   </div> 
   <button className="button" onClick={props.onRemove}>Confirm</button>
 </Modal>
);

export default ExpenseModal;