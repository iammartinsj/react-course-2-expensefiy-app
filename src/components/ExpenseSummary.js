import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesCount from '../selectors/expenses-count';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00');
  return (
    <div>
      <h2>Viewing {expensesCount} {expenseWord} totalling {formattedExpensesTotal}</h2>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filter);

  return {
    expensesCount: selectExpensesCount(visibleExpenses),
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }
};

export default connect(mapStateToProps)(ExpensesSummary);