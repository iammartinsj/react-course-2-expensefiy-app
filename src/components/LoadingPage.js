import React from 'react';

export default () => (
  <div className="loader">
    <img className="loader__spinner" src="./images/loader.gif" />
  </div>
);