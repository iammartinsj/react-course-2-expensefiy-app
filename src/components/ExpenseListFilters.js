import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setFilterText, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  onTextChange = (e) => {
    this.props.setFilterText(e.target.value);
  }

  onSortByChange = (e) => {
    if(e.target.value === 'date') {
      this.props.sortByDate() 
    } else if(e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  }

  render() {
    return (
      <div>
        <input 
          type="text" 
          value={this.props.filters.text} 
          onChange={this.onTextChange} 
        />
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortByChange}
        > 
          <option value="amount">Amount</option>
          <option value="date">Date</option> 
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters:state.filter
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)), 
  setFilterText: (text) => dispatch(setFilterText(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
