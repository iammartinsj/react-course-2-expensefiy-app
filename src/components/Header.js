import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({startLogout}) => (
  <div>
    <h1>
       Expensify
    </h1>
    <ul>
      <li><NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink></li>
      <li><NavLink to="/create" activeClassName="is-active">Create Expense</NavLink></li>
      <li><NavLink to="/help" activeClassName="is-active">Help</NavLink></li>
      <li><button onClick={startLogout}>Logout</button></li>
    </ul>   
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);