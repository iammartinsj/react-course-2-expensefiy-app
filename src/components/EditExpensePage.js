import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseModal from './ExpenseModal';

export class EditExpensePage extends React.Component {
  state = {
    forRemovalExpense: undefined
  }

  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onSetForRemoval = () => {
    this.setState(() => ({ 
      forRemovalExpense: this.props.expense.description 
    }));
  };

  onClearForRemovalExpense = () => {
    this.setState(() => ({
      forRemovalExpense: undefined
    }));
  }

  onRemove = () => {
    this.setState(() => ({
      forRemovalExpense: undefined
    }));
    this.props.startRemoveExpense({id:this.props.expense.id});
    this.props.history.push('/');
  }

  render(){
    return(
      <div>
        <div className="page-header" >
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm 
            expense={this.props.expense} 
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.onSetForRemoval}>
            Remove Expense
          </button>
        </div>
        <ExpenseModal 
          forRemovalExpense={this.state.forRemovalExpense}
          onRemove={this.onRemove}
          onClearForRemovalExpense={this.onClearForRemovalExpense}
        />
      </div>
    )
  };
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);