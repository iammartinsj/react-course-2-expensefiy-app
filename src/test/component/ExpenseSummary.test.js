import React from 'react';
import { shallow } from 'enzyme'; 
import { ExpensesSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';
import expensesCount from '../../selectors/expenses-count';
import expensesTotal from '../../selectors/expenses-total';

test('should render one Expense Summary Correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={500}  />);
  expect(wrapper).toMatchSnapshot();
});

test('should render multiple Expenses Summary Correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={3} expenseTotal={1500} />);
  expect(wrapper).toMatchSnapshot();
});