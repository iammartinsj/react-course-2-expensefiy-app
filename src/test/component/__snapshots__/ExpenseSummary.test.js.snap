// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render multiple Expenses Summary Correctly 1`] = `
<div>
  <h2>
    Viewing 
     
    expenses
     totalling 
    ₱0.00
  </h2>
</div>
`;

exports[`should render one Expense Summary Correctly 1`] = `
<div>
  <h2>
    Viewing 
     
    expenses
     totalling 
    ₱0.00
  </h2>
</div>
`;
